File: Longest Palindromic Substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int N = s.length();
        int maxlen = 1;
        int arg_i = 0;
        int arg_j = 0;
        for (int i = 0; i < N; i++) {
            //center at i
            int len = 0;
            while (i - len >= 0 && i + len < N) {
                if (s[i-len] == s[i+len]) {
                    if (2 * len + 1 > maxlen) {
                        maxlen = 2 * len + 1;
                        arg_i = i - len;
                        arg_j = i + len;
                    }
                }
                else
                    break;
                len++;
            }
            
            //center at i and i+1
            if (i < N-1 && s[i] == s[i+1]) {
                len = 0;
                while (i - len >= 0 && i + 1 + len < N) {
                    if (s[i-len] == s[i+1+len]) {
                        if (2 * len + 2 > maxlen) {
                            maxlen = 2 * len + 2;
                            arg_i = i - len;
                            arg_j = i + 1 + len;
                        }
                    }
                    else
                        break;
                    len++;
                }
            }
        }
        return s.substr(arg_i, arg_j - arg_i + 1);
    }
};

File: Longest Palindromic Substring.py
class Solution:
	# @return a string
	def longestPalindrome(self, s):
		N = len(s)
		maxsz = 0
		argi = 0
		argj = 0
		# zigzag = N/2, N/2-1, N/2+1, N/2-2, N/2+2...
		zigzag = [N/2]
		for i in range(1,N/2+1):
			zigzag.append(N/2-i)
			zigzag.append(N/2+i)
		if (zigzag[len(zigzag)-1] == N):
			zigzag = zigzag[:-1]
			
		for i in zigzag:
			# maxlen would be 2*i+2 or 2*(N-1-i)+1
			if (2*i+2 < maxsz):
				continue
			if (2*(N-i-1)+1 < maxsz):
				continue
			for j in [0,1]:
				sz = 0
				while (i-sz >= 0 and i+j+sz < N):
					if (s[i-sz] == s[i+j+sz]):
						if (2*sz+j > maxsz):
							maxsz = 2*sz+j
							argi = i-sz
							argj = i+j+sz
					else:
						break
					sz = sz + 1
		return s[argi:argj+1]